### PR TITLE
For #1169: Use wild-card type parameters in ctor

### PR DESCRIPTION
**Pull request by andreoss**
Per #1169:

- Make ctors accept wild-carded types

### CODE REVIEW

**Comment by victornoel** [comment]
@0crat status

**Comment by victornoel** [comment]
@0crat status

**Comment by fabriciofx** [comment]
@0crat status

**Comment by 0crat** [comment]
> @0crat status ([here](https://github.com/yegor256/cactoos/issues/1477#issuecomment-701521096))

@victornoel This is what I know about this job in C63314D6Z, as in [§32](http://www.zerocracy.com/policy.html#32):

  * The job [#1477](https://github.com/yegor256/cactoos/issues/1477) is [in scope](http://www.zerocracy.com/policy.html#14) for 2days
  * The role is `REV`
  * The job is assigned to @fabriciofx[/z](https://www.0crat.com/u/fabriciofx) for [2days](http://www.zerocracy.com/policy.html#8)
  * There is no monetary reward attached, it's a [free](http://www.zerocracy.com/policy.html#2) job
  * The job doesn't have any [impediments](http://www.zerocracy.com/policy.html#9)
  * The [budget](http://www.zerocracy.com/policy.html#4) is 15 minutes/points
  * These users are banned and won't be assigned:
    * @andreoss[/z](https://www.0crat.com/u/andreoss): This user reported the ticket
  * Job [footprint](https://www.0crat.com/footprint/C63314D6Z?q=%7Bjob%3A%27gh%3Ayegor256%2Fcactoos%231477%27%2Ctype%3A%7B%24not%3A%2F%5ENotify%2F%7D%7D) (restricted area)

**Comment by 0crat** [comment]
> @0crat status ([here](https://github.com/yegor256/cactoos/issues/1477#issuecomment-701662150))

@fabriciofx This is what I know about this job in C63314D6Z, as in [§32](http://www.zerocracy.com/policy.html#32):

  * The job [#1477](https://github.com/yegor256/cactoos/issues/1477) is [in scope](http://www.zerocracy.com/policy.html#14) for 2days
  * The role is `REV`
  * The job is assigned to @fabriciofx[/z](https://www.0crat.com/u/fabriciofx) for [2days](http://www.zerocracy.com/policy.html#8)
  * There is no monetary reward attached, it's a [free](http://www.zerocracy.com/policy.html#2) job
  * The job doesn't have any [impediments](http://www.zerocracy.com/policy.html#9)
  * The [budget](http://www.zerocracy.com/policy.html#4) is 15 minutes/points
  * These users are banned and won't be assigned:
    * @andreoss[/z](https://www.0crat.com/u/andreoss): This user reported the ticket
  * Job [footprint](https://www.0crat.com/footprint/C63314D6Z?q=%7Bjob%3A%27gh%3Ayegor256%2Fcactoos%231477%27%2Ctype%3A%7B%24not%3A%2F%5ENotify%2F%7D%7D) (restricted area)

**Comment by 0crat** [comment]
> @0crat status ([here](https://github.com/yegor256/cactoos/issues/1477#issuecomment-701612528))

@victornoel This is what I know about this job in C63314D6Z, as in [§32](http://www.zerocracy.com/policy.html#32):

  * The job [#1477](https://github.com/yegor256/cactoos/issues/1477) is [in scope](http://www.zerocracy.com/policy.html#14) for 2days
  * The role is `REV`
  * The job is assigned to @fabriciofx[/z](https://www.0crat.com/u/fabriciofx) for [2days](http://www.zerocracy.com/policy.html#8)
  * There is no monetary reward attached, it's a [free](http://www.zerocracy.com/policy.html#2) job
  * The job doesn't have any [impediments](http://www.zerocracy.com/policy.html#9)
  * The [budget](http://www.zerocracy.com/policy.html#4) is 15 minutes/points
  * These users are banned and won't be assigned:
    * @andreoss[/z](https://www.0crat.com/u/andreoss): This user reported the ticket
  * Job [footprint](https://www.0crat.com/footprint/C63314D6Z?q=%7Bjob%3A%27gh%3Ayegor256%2Fcactoos%231477%27%2Ctype%3A%7B%24not%3A%2F%5ENotify%2F%7D%7D) (restricted area)

File: src/main/java/org/cactoos/iterable/IterableOf.java [hunk 1]
```diff
@@ -76,21 +77,21 @@ public IterableOf(final List<X> list) {
...
      */
-    public IterableOf(final Scalar<Iterator<X>> sclr) {
+    public IterableOf(final Scalar<Iterator<? extends X>> sclr) {
         this.itr = sclr;
     }
 
     @Override
     public Iterator<X> iterator() {
-        return new Unchecked<>(this.itr).value();
+        return new Mapped<>(x -> x, new Unchecked<>(this.itr).value());
```
Comment (fabriciofx) [inline src/main/java/org/cactoos/iterable/IterableOf.java:94]: @andreoss Has been needed put a `Mapped` here? Why? An `IteratorOf` can be used instead?

File: src/main/java/org/cactoos/iterable/Matched.java [hunk 2]
```diff
@@ -54,7 +54,9 @@
      * @param fst The first part of duplex iterator.
      * @param snd The second part of duplex iterator.
      */
-    public Matched(final Iterable<X> fst, final Iterable<X> snd) {
+    public Matched(
+        final Iterable<? extends X> fst, final Iterable<? extends X> snd
```
Comment (fabriciofx) [inline src/main/java/org/cactoos/iterable/Matched.java:58]: @andreoss Please, break the line here, putting one parameter for line.

File: src/main/java/org/cactoos/iterable/Matched.java [hunk 3]
```diff
@@ -66,12 +68,12 @@ public Matched(final Iterable<X> fst, final Iterable<X> snd) {
      */
     public Matched(
         final BiFunc<X, X, Boolean> fnc,
-        final Iterable<X> fst, final Iterable<X> snd
+        final Iterable<? extends X> fst, final Iterable<? extends X> snd
```
Comment (fabriciofx) [inline src/main/java/org/cactoos/iterable/Matched.java:71]: @andreoss the same above.

File: src/main/java/org/cactoos/iterable/Solid.java [hunk 4]
```diff
@@ -46,8 +46,8 @@ public Solid(final X... src) {
      * Ctor.
      * @param iterable The iterable
      */
-    public Solid(final Iterable<X> iterable) {
-        super(new Synced<>(new Sticky<>(iterable)));
+    public Solid(final Iterable<? extends X> iterable) {
+        super(new Synced<>(new Sticky<>(new Mapped<>(x -> x, iterable))));
```
Comment (fabriciofx) [inline src/main/java/org/cactoos/iterable/Solid.java:50]: @andreoss the same above: why `Mapped`?

File: src/main/java/org/cactoos/iterator/Filtered.java [hunk 5]
```diff
@@ -75,7 +75,9 @@
      * @param fnc Predicate
      * @param src Source iterable
      */
-    public Filtered(final Func<X, Boolean> fnc, final Iterator<X> src) {
+    public Filtered(
+        final Func<X, Boolean> fnc, final Iterator<? extends X> src
```
Comment (fabriciofx) [inline src/main/java/org/cactoos/iterator/Filtered.java:79]: @andreoss the same above.

File: src/main/java/org/cactoos/iterator/TailOf.java [hunk 6]
```diff
@@ -41,16 +41,19 @@
...
      */
-    public TailOf(final int num, final Iterator<T> iterator) {
+    public TailOf(final int num, final Iterator<? extends T> iterator) {
         super(
-            new Reversed<>(
-                new HeadOf<>(
-                    num,
-                    new Reversed<>(
-                        new IterableOf<>(iterator)
+            new Mapped<>(
```
Comment (fabriciofx) [inline src/main/java/org/cactoos/iterator/TailOf.java:46]: @andreoss the same above.

File: src/main/java/org/cactoos/map/Synced.java [hunk 7]
```diff
@@ -121,16 +125,18 @@ public Synced(final Map<X, Y> map, final Map.Entry<X, Y>... list) {
      * @param list List of items
      * @param <Z> Type of items in the list
      */
-    public <Z> Synced(final Func<Z, Map.Entry<X, Y>> entry,
-        final Map<X, Y> map, final Iterable<Z> list) {
+    public <Z> Synced(
+        final Func<Z, Map.Entry<? extends X, ? extends Y>> entry,
+        final Map<X, Y> map, final Iterable<Z> list
```
Comment (fabriciofx) [inline src/main/java/org/cactoos/map/Synced.java:130]: @andreoss the same above.

File: src/test/java/org/cactoos/list/JoinedTest.java [hunk 8]
```diff
@@ -435,4 +439,25 @@ public void itemAndList() {
             )
         ).affirm();
     }
+
+    @Test
+    public void infersCorrectly() {
```
Comment (fabriciofx) [inline src/test/java/org/cactoos/list/JoinedTest.java:444]: @andreoss please, add more tests, for a `map` and the decorators where were changed.

**Review by fabriciofx — CHANGES_REQUESTED** [review]
@andreoss please, take a look.

File: src/main/java/org/cactoos/iterable/IterableOf.java (hunk 1 above)
Comment (andreoss) [reply src/main/java/org/cactoos/iterable/IterableOf.java:94]: @fabriciofx `Mapped` is used in order to coerce `? extends X` to `X`. `IteratorOf` wouldn't be enough for that

File: src/main/java/org/cactoos/iterator/Filtered.java (hunk 5 above)
Comment (andreoss) [reply src/main/java/org/cactoos/iterator/Filtered.java:79]: @fabriciofx Fixed

File: src/main/java/org/cactoos/iterable/Matched.java (hunk 2 above)
Comment (andreoss) [reply src/main/java/org/cactoos/iterable/Matched.java:58]: @fabriciofx fixed

File: src/main/java/org/cactoos/map/Synced.java (hunk 7 above)
Comment (andreoss) [reply src/main/java/org/cactoos/map/Synced.java:130]: @fabriciofx Fixed

**Comment by andreoss** [comment]
@fabriciofx `Mapped` is used in order to turn `? extends X` to  `X`.  I'm not sure of `IteratorOf` or `IterableOf` should have such feature in their ctors. 
Also, it seems that the tests are not necessary for these changes, since they only can cause a compile-time failures.

**Comment by victornoel** [comment]
@fabriciofx @andreoss about this thing with `Mapped`, I agree it's not great to have to go trough the use of `Mapped` just because of some java limitation about typing. I'm not sure if there is a better way, but I'm going to think a bit about it too on my side :)

**Comment by victornoel** [comment]
@fabriciofx @andreoss where are we with this PR? :)

**Comment by andreoss** [comment]
@victornoel Do you think we need more tests here, or that's complier's work?

**Comment by victornoel** [comment]
@andreoss I haven't read everything in details yet, but I think the only advantages of tests would be to avoid regressions in future changes of the lib. Something like acceptance tests... 

Maybe we can add a todo for it or maybe start on another ticket a new stream of work around having a strong acceptance test suite, I'm not sure. It's true that this change could be tested a bit. Let's add a todo then and it should be clear what is to be tested and that the tests must be clear about that also so that we remember why they were introduced, WDYT?

**Comment by andreoss** [comment]
@victornoel I would add a todo for dealing with `Mapped`  issue. It could be replaced by unsafe casting

File: src/main/java/org/cactoos/iterable/Filtered.java [hunk 9]
```diff
@@ -65,7 +65,7 @@ public Filtered(final Func<X, Boolean> fnc, final X... src) {
      * @param fnc Predicate
      * @param src Source iterable
      */
-    public Filtered(final Func<X, Boolean> fnc, final Iterable<X> src) {
+    public Filtered(final Func<X, Boolean> fnc, final Iterable<? extends X> src) {
```
Comment (victornoel) [inline src/main/java/org/cactoos/iterable/Filtered.java:68]: @andreoss to be complete, we need `Func<? super X, Boolean>`, so that we can write for example:
```java
 Iterable<Double> res = new Filtered<Double>((Number d) -> d.doubleValue() > 0, new IterableOf<Double>(4.1));
```

That's why I guess we need a few tests..

File: src/main/java/org/cactoos/iterable/Mapped.java [hunk 10]
```diff
@@ -51,7 +51,7 @@ public Mapped(final Func<X, Y> fnc, final X... src) {
      * @param fnc Func
      * @param src Source iterable
      */
-    public Mapped(final Func<X, Y> fnc, final Iterable<X> src) {
+    public Mapped(final Func<X, Y> fnc, final Iterable<? extends X> src) {
```
Comment (victornoel) [inline src/main/java/org/cactoos/iterable/Mapped.java:54]: @andreoss here I think we need:
- to move the the `X` argument at the constructor level, it's not needed at the class level IMHO
- have `Func<? super X, ? extends Y>` to be able to express more things

File: src/main/java/org/cactoos/iterable/Matched.java [hunk 11]
```diff
@@ -66,12 +69,13 @@ public Matched(final Iterable<X> fst, final Iterable<X> snd) {
      */
     public Matched(
         final BiFunc<X, X, Boolean> fnc,
```
Comment (victornoel) [inline src/main/java/org/cactoos/iterable/Matched.java:71]: @andreoss here I think it should be `BiFunc<? super X, ? super X, Boolean>`

**Review by victornoel — CHANGES_REQUESTED** [review]
@andreoss I've added a few comments about the types of the func or equivalent passed along the iterables in those class constructors. I didn't go though the whole PR (it's long, btw next time, let's try to focus on a smaller area of the codebase :) but I think you can get the idea about having function applied to the element of the iterable to be `? super X`.

File: src/main/java/org/cactoos/iterable/Mapped.java (hunk 10 above)
Comment (andreoss) [reply src/main/java/org/cactoos/iterable/Mapped.java]: @victornoel Added todo for removing `X` as a class level type parameter

File: src/main/java/org/cactoos/iterable/Matched.java (hunk 11 above)
Comment (andreoss) [reply src/main/java/org/cactoos/iterable/Matched.java]: @victornoel Fixed

File: src/main/java/org/cactoos/iterable/Mapped.java [hunk 12]
```diff
@@ -33,6 +33,10 @@
  * @param <X> Type of source item
  * @param <Y> Type of target item
  * @since 0.1
+ * @todo #1169:30m Remove type parameter X from this
+ *  class definition, and put extra type parameter on each ctor.
+ *  Possibly, the same change is possible in iterator.Mapped,
```
Comment (victornoel) [inline src/main/java/org/cactoos/iterable/Mapped.java:38]: @andreoss not possibly, this should be done too :) also IMHO there is no reason for those two tasks to be done as once, just say in the todo what is to be done, and then the assignee will decide if everything can be done at once.

File: src/main/java/org/cactoos/iterable/Mapped.java [hunk 13]
```diff
@@ -42,7 +46,7 @@
      * @param src Source iterable
      */
     @SafeVarargs
-    public Mapped(final Func<X, Y> fnc, final X... src) {
+    public Mapped(final Func<? super X, Y> fnc, final X... src) {
```
Comment (victornoel) [inline src/main/java/org/cactoos/iterable/Mapped.java:49]: @andreoss it's missing the `? extend Y` in the second func type parameter.

File: src/main/java/org/cactoos/iterable/Filtered.java [hunk 14]
```diff
@@ -78,7 +78,7 @@ public Filtered(final Func<X, Boolean> fnc, final Iterable<X> src) {
      * @param src Source iterable
      * @param fnc Predicate
      */
-    public Filtered(final Iterable<X> src, final Func<X, Scalar<Boolean>> fnc) {
+    public Filtered(final Iterable<X> src, final Func<? super X, Scalar<Boolean>> fnc) {
```
Comment (victornoel) [inline src/main/java/org/cactoos/iterable/Filtered.java:81]: @andreoss we are missing the `? extends X` on the src parameter here

File: src/main/java/org/cactoos/iterable/Matched.java [hunk 15]
```diff
@@ -65,13 +68,14 @@ public Matched(final Iterable<X> fst, final Iterable<X> snd) {
      * @param snd The second part of duplex iterator.
      */
     public Matched(
-        final BiFunc<X, X, Boolean> fnc,
-        final Iterable<X> fst, final Iterable<X> snd
+        final BiFunc<? super X, ? super X, Boolean> fnc,
+        final Iterable<? extends X> fst,
+        final Iterable<? extends X> snd
```
Comment (victornoel) [inline src/main/java/org/cactoos/iterable/Matched.java:73]: @andreoss let's replace the type of `snd` with `Iterable<? extends Y>` with Y being a type declared at the constructor level (or you can just add a todo). There is no need for this second parameter to be so constrained in types.

File: src/main/java/org/cactoos/iterator/Mapped.java [hunk 16]
```diff
@@ -53,7 +53,7 @@
      * @param func Func
      * @param iterator Source iterator
      */
-    public Mapped(final Func<X, Y> func, final Iterator<X> iterator) {
+    public Mapped(final Func<X, Y> func, final Iterator<? extends X> iterator) {
```
Comment (victornoel) [inline src/main/java/org/cactoos/iterator/Mapped.java:56]: @andreoss here too let's apply the same thing with super and extends on the func

File: src/main/java/org/cactoos/map/MapOf.java [hunk 17]
```diff
@@ -106,8 +104,10 @@ public MapOf(final Map<X, Y> src, final Map.Entry<X, Y>... list) {
      * @param <Z> Type of items in the list
      * @since 0.11
      */
-    public <Z> MapOf(final Func<Z, Map.Entry<X, Y>> entry,
-        final Iterable<Z> list) {
+    public <Z> MapOf(
+        final Func<Z, Map.Entry<? extends X, ? extends Y>> entry,
```
Comment (victornoel) [inline src/main/java/org/cactoos/map/MapOf.java:108]: @andreoss I believe the func should be taking a `? super Z` no?

File: src/main/java/org/cactoos/map/Synced.java [hunk 18]
```diff
@@ -70,9 +70,12 @@ public Synced(final Map<X, Y> map, final Map.Entry<X, Y>... list) {
      * @param <Z> Type of items in the list
      * @checkstyle ParameterNumberCheck (5 lines)
      */
-    public <Z> Synced(final Func<Z, X> key,
-        final Func<Z, Y> value, final Map<X, Y> map,
-        final Iterable<Z> list) {
+    public <Z> Synced(
+        final Func<Z, X> key,
+        final Func<Z, Y> value,
```
Comment (victornoel) [inline src/main/java/org/cactoos/map/Synced.java:75]: @andreoss here too let's apply the same concepts

File: src/main/java/org/cactoos/set/Sorted.java [hunk 19]
```diff
@@ -54,7 +54,10 @@ public Sorted(final Comparator<T> cmp, final T... array) {
      * @param src An {@link Iterable}
      */
     @SuppressWarnings("PMD.ConstructorOnlyInitializesOrCallOtherConstructors")
-    public Sorted(final Comparator<T> cmp, final Iterable<T> src) {
+    public Sorted(
+        final Comparator<T> cmp,
```
Comment (victornoel) [inline src/main/java/org/cactoos/set/Sorted.java:58]: @andreoss I believe the comparator should take a `? super T` here too (a comparator is basically a func from T,T to bool :)

File: src/main/java/org/cactoos/iterable/IterableOf.java (hunk 1 above)
Comment (victornoel) [reply src/main/java/org/cactoos/iterable/IterableOf.java:94]: @andreoss please add a todo so that this can be investigated and the code simplified, there must be some way to do something simpler.

**Review by victornoel — CHANGES_REQUESTED** [review]
@andreoss as you can see, there is still a lot to change still :) either go for it, or create a clear todo about adding the wildcard types to all the function and similar.

File: src/main/java/org/cactoos/iterable/Filtered.java (hunk 9 above)
Comment (andreoss) [reply src/main/java/org/cactoos/iterable/Filtered.java]: @victornoel Added a test

File: src/main/java/org/cactoos/iterable/Mapped.java (hunk 10 above)
Comment (andreoss) [reply src/main/java/org/cactoos/iterable/Mapped.java]: @victornoel Actually, I've done it in this PR.

File: src/main/java/org/cactoos/map/Synced.java (hunk 18 above)
Comment (andreoss) [reply src/main/java/org/cactoos/map/Synced.java]: @victornoel Done

File: src/main/java/org/cactoos/map/MapOf.java (hunk 17 above)
Comment (andreoss) [reply src/main/java/org/cactoos/map/MapOf.java]: @victornoel Done.

File: src/main/java/org/cactoos/iterable/Matched.java (hunk 15 above)
Comment (andreoss) [reply src/main/java/org/cactoos/iterable/Matched.java]: @victornoel Done

File: src/main/java/org/cactoos/iterable/Filtered.java (hunk 14 above)
Comment (andreoss) [reply src/main/java/org/cactoos/iterable/Filtered.java:81]: @victornoel Done

File: src/main/java/org/cactoos/iterable/Mapped.java (hunk 13 above)
Comment (andreoss) [reply src/main/java/org/cactoos/iterable/Mapped.java]: @victornoel Done

File: src/main/java/org/cactoos/iterator/Mapped.java (hunk 16 above)
Comment (andreoss) [reply src/main/java/org/cactoos/iterator/Mapped.java]: @victornoel Done

**Comment by andreoss** [comment]
@fabriciofx @victornoel Please take a look once more.

File: src/main/java/org/cactoos/iterable/IterableOf.java [hunk 20]
```diff
@@ -45,14 +45,16 @@
  * @param <X> Type of item
  * @since 0.12
  * @checkstyle ClassDataAbstractionCouplingCheck (550 lines)
+ * @todo #1169:30m This class uses Mapped in order to coerce wild-card type.
+ *  Research a more eloquent solution and remove usage of Mapped.
```
Comment (victornoel) [inline src/main/java/org/cactoos/iterable/IterableOf.java:49]: @andreoss I see you replaced the usage of Mapped with a cast, so this todo should be updated I think. Actually, I'm wondering if casting is not simply acceptable here… WDYT?

File: src/main/java/org/cactoos/iterable/Solid.java (hunk 4 above)
Comment (victornoel) [inline src/main/java/org/cactoos/iterable/Solid.java:50]: @andreoss here we have mapped, we could either do casting as discussed above or add a todo to find a better solution. But I would prefer that if we choose one way, we apply the same everywhere though.

File: src/main/java/org/cactoos/iterable/Synced.java [hunk 21]
```diff
@@ -73,15 +77,15 @@ public Synced(final Iterable<X> iterable) {
...
+    public Synced(final Object lck, final Iterable<? extends X> iterable) {
         this.origin = iterable;
         this.lock = lck;
     }
 
     @Override
     public Iterator<X> iterator() {
         synchronized (this.lock) {
-            return this.origin.iterator();
+            return new Mapped<>(x -> x, this.origin.iterator());
```
Comment (victornoel) [inline src/main/java/org/cactoos/iterable/Synced.java:88]: @andreoss same here about cast vs mapped, and also about the todo

File: src/main/java/org/cactoos/iterator/Mapped.java [hunk 22]
```diff
@@ -24,53 +24,59 @@
...
+                @Override
+                public Y next() {
+                    if (this.hasNext()) {
+                        return new Unchecked<>(
+                            new org.cactoos.scalar.Mapped<>(
+                                func, iterator::next
+                            )
+                        ).value();
+                    }
+                    throw new NoSuchElementException();
```
Comment (victornoel) [inline src/main/java/org/cactoos/iterator/Mapped.java:70]: @andreoss why change the implementation of `next()`? Wasn't the previous one correct?

File: src/main/java/org/cactoos/map/Synced.java [hunk 23]
```diff
@@ -139,16 +156,19 @@ public Synced(final Iterable<Map.Entry<X, Y>> list) {
...
 
     /**
      * Ctor.
      * @param map The map
      */
-    public Synced(final Map<X, Y> map) {
-        super(Collections.synchronizedMap(map));
+    @SuppressWarnings("unchecked")
+    public Synced(final Map<? extends X, ? extends Y> map) {
+        super(Collections.synchronizedMap(new MapOf<>(map)));
```
Comment (victornoel) [inline src/main/java/org/cactoos/map/Synced.java:172]: @andreoss here we cannot use `MapOf` because it would create a new map instead of making the wrapped one synchronized.

File: src/main/java/org/cactoos/set/Sorted.java (hunk 19 above)
Comment (victornoel) [inline src/main/java/org/cactoos/set/Sorted.java:58]: @andreoss here there is still one comparator without the `? super T`

File: src/test/java/org/cactoos/iterable/MappedTest.java [hunk 24]
```diff
@@ -43,10 +42,11 @@
 final class MappedTest {
 
     @Test
+    @SuppressWarnings("unchecked")
```
Comment (victornoel) [inline src/test/java/org/cactoos/iterable/MappedTest.java:45]: @andreoss why is this needed?

File: src/test/java/org/cactoos/list/JoinedTest.java [hunk 25]
```diff
@@ -435,4 +439,25 @@ public void itemAndList() {
...
+        new Assertion<>(
+            "must be able to infer type of elements",
```
Comment (victornoel) [inline src/test/java/org/cactoos/list/JoinedTest.java:446]: @andreoss while we are at it, let's put an upper case at the beginning of the sentence

**Review by victornoel — CHANGES_REQUESTED** [review]
@andreoss sorry, some more comments :) if this starts to become out of hand, don't hesitate to recreate a smaller PR, but if you are good with it, let's go all the way :)

File: src/main/java/org/cactoos/iterator/Mapped.java (hunk 22 above)
Comment (andreoss) [reply src/main/java/org/cactoos/iterator/Mapped.java:70]: @victornoel I wanted to reuse `scalar.Mapped` here, to be consistent between different Mapped implementations. 
FindBugs was triggered by the lack of a `hasNext` check, not sure why it used to accept  the old code

File: src/main/java/org/cactoos/set/Sorted.java (hunk 19 above)
Comment (andreoss) [reply src/main/java/org/cactoos/set/Sorted.java:58]: @victornoel Done

File: src/test/java/org/cactoos/map/GroupedTest.java [hunk 26]
```diff
@@ -83,4 +83,22 @@ void groupedByOneHasEntries() throws Exception {
             )
         );
     }
+
+    @Test
+    void groupedBySuperType() throws Exception {
+        MatcherAssert.assertThat(
```
[the remark this answers is not in the document]
Comment (andreoss) [reply src/test/java/org/cactoos/map/GroupedTest.java]: @victornoel Done

File: src/test/java/org/cactoos/list/JoinedTest.java (hunk 25 above)
Comment (andreoss) [reply src/test/java/org/cactoos/list/JoinedTest.java]: @victornoel Done

File: src/main/java/org/cactoos/map/Synced.java (hunk 23 above)
Comment (andreoss) [reply src/main/java/org/cactoos/map/Synced.java]: @victornoel Done.

File: src/main/java/org/cactoos/iterable/Synced.java (hunk 21 above)
Comment (andreoss) [reply src/main/java/org/cactoos/iterable/Synced.java]: @victornoel Replaced with casting.

File: src/main/java/org/cactoos/iterable/Solid.java (hunk 4 above)
Comment (andreoss) [reply src/main/java/org/cactoos/iterable/Solid.java]: @victornoel Here it was unnecessary. Removed it

File: src/main/java/org/cactoos/iterable/IterableOf.java (hunk 20 above)
Comment (andreoss) [reply src/main/java/org/cactoos/iterable/IterableOf.java]: @victornoel Removed. It seems there is no a better solution than casting for that (which could be put a decorator)

File: src/main/java/org/cactoos/iterator/Mapped.java (hunk 22 above)
Comment (victornoel) [reply src/main/java/org/cactoos/iterator/Mapped.java:70]: @andreoss I see, personally I find that `scalar.Mapped` makes things harder to read but it is correct anyway, let's keep it like you wrote it :)

File: src/main/java/org/cactoos/scalar/Mapped.java [hunk 27]
```diff
@@ -28,45 +28,34 @@
...
-        this.origin = scalar;
-        this.func = mapping;
-    }
-
-    @Override
-    public U value() throws Exception {
-        return this.func.apply(this.origin.value());
+    public <T> Mapped(
+        final Func<? super T, ? extends U> func,
+        final Scalar<? extends T> scalar
```
Comment (victornoel) [inline src/main/java/org/cactoos/scalar/Mapped.java:56]: @andreoss  seeing this change, I wonder if the next step wouldn't be to do the same thing with covariance with scalars? Do you have an opinion on this?

File: src/main/java/org/cactoos/set/Sorted.java (hunk 19 above)
Comment (victornoel) [reply src/main/java/org/cactoos/set/Sorted.java:58]: @andreoss I still don't see it in the PR's diff...

**Review by victornoel — CHANGES_REQUESTED** [review]
@andreoss there seem to still be one little thing to change, see my comment.

Also, about `Assertion`; it wasn't needed to change existing code, only the new code. There exists a ticket for doing that job, in the future please refrain from doing too much, it makes the PR harder to review :) ok?

**Comment by victornoel** [comment]
@andreoss also it seems that coverage has been reduced on `iterator.Mapped`, if something easy can be done about it it would be great so that all the checks pass.

File: src/main/java/org/cactoos/scalar/Mapped.java (hunk 27 above)
Comment (andreoss) [reply src/main/java/org/cactoos/scalar/Mapped.java:56]: @victornoel Yes, `func` and `proc` packages as well

Comment (victornoel) [reply src/main/java/org/cactoos/scalar/Mapped.java:56]: @andreoss thx, I will create a new ticket in the future, because I have to think about it a bit before :)

**Comment by andreoss** [comment]
@victornoel 
Added wild cards to `set.Sorted`
Fixed coverage check

Please take a look again

**Review by victornoel — APPROVED** [review]
@andreoss thanks :) let's try to keep PRs smaller next time, but this is great work anywa :)

**Comment by victornoel** [comment]
@rultor merge

**Comment by 0crat** [comment]
Code review was too long (24 days), architects (@victornoel) were penalized, see [§55](http://www.zerocracy.com/policy.html#55)

**Comment by 0crat** [comment]
@sereshqua[/z](https://www.0crat.com/u/sereshqua) please review this job completed by @fabriciofx[/z](https://www.0crat.com/u/fabriciofx), as in [§30](http://www.zerocracy.com/policy.html#30); the job will be fully closed and all payments will be made when the quality review is completed

**Comment by sereshqua** [comment]
@0crat quality good
